napari/_qt/qt_main_window.py

```diff
@@ -635,6 +642,52 @@ def remove_dock_widget(self, widget: QWidget):
         # rather than directly using _add_viewer_dock_widget
         _dw.deleteLater()
 
+    def add_function_widget(
+        self,
+        function,
+        *,
+        magic_kwargs=None,
+        name: str = '',
+        area: str = 'bottom',
+        allowed_areas=None,
+        shortcut=None,
+    ):
+        """Turn a function into a dock widget via magicgui.
+
+        Parameters
+        ----------
+        function : callable
+            Function that you want to add.
+        magic_kwargs : dict, optional
+            Keyword arguments to :func:`magicgui.magicgui` that
+            can be used to specify widget.
+        name : str, optional
+            Name of dock widget to appear in window menu.
+        area : str
+            Side of the main window to which the new dock widget will be added.
+            Must be in {'left', 'right', 'top', 'bottom'}
+        allowed_areas : list[str], optional
+            Areas, relative to main window, that the widget is allowed dock.
+            Each item in list must be in {'left', 'right', 'top', 'bottom'}
+            By default, all areas are allowed.
+        shortcut : str, optional
+            Keyboard shortcut to appear in dropdown menu.
+
+        Returns
+        -------
+        dock_widget : QtViewerDockWidget
+            `dock_widget` that can pass viewer events.
+        """
+        from magicgui import magicgui
+
+        return self.add_dock_widget(
+            magicgui(**magic_kwargs or {})(function),
```

Comment:
one more suggestion: when not using as a decorator, this syntax is probably clearer:
```suggestion
            magicgui(function, **magic_kwargs or {}),
```